readthedocs/proxito/views/mixins.py

```diff
@@ -30,52 +36,129 @@ class ServeDocsMixin:
 
     version_type = INTERNAL
 
-    def _serve_docs(
-            self,
-            request,
-            final_project,
-            path,
+    def _serve_docs(self, request, project, version, filename, check_if_exists=False):
+        """
+        Serve a documentation file.
+
+        :param check_if_exists: If `True` we check if the file exists before trying
+         to serve it.
+        """
+        base_storage_path = project.get_storage_path(
+            type_=MEDIA_TYPE_HTML,
+            version_slug=version.slug,
+            include_file=False,
+            version_type=version.type,
+        )
+
+        # Handle our backend storage not supporting directory indexes,
+        # so we need to append index.html when appropriate.
+        if not filename or filename.endswith("/"):
+            filename += "index.html"
+
+        # If the filename starts with `/`, the join will fail,
+        # so we strip it before joining it.
+        storage_path = build_media_storage.join(base_storage_path, filename.lstrip("/"))
+
+        if check_if_exists and not build_media_storage.exists(storage_path):
+            return None
```

Comment:
I may be clearer to `raise StorageFileNotExists` or similar and handle this exception from where it's called. That will be easier to read and follow than understanding what `None` means in that context.